zeek/Shaping-Zeek-NDJSON.md

```diff
@@ -0,0 +1,210 @@
+
+   * The data type of each field in the type definition is applied to the
+     field of that name in the input record. (Note: This could be performed
+     separately via the `cast()` function.)
+
+   * The fields in the input record are ordered to match the order in which
+     they appear in the type definition. (Note: This could be performed
+     separately via the `order()` function.)
+
+   Any fields that appear in the input record that are not present in the
+   type definition are kept and assigned an inferred data type. If you would
+   prefer to have such additional fields dropped (i.e., to maintain strict
+   adherence to the shape), append a call to the `crop()` function to the
+   Zed pipeline, e.g.:
+
+      ```
+      put this := unflatten(this) | put this := shape(schemas[_path]) | put this := crop(schemas[_path])
+      ```
+
+   Open issues [zed/2585](https://github.com/brimdata/zed/issues/2585) and
+   [zed/2776](https://github.com/brimdata/zed/issues/2776) both track planned
+   future improvements to this part of the configuration.
+
+# Invoking the Shaper From `zq`
+
+A shaper is typically invoked via the `-I` option of `zq`.
+
+For example, if working in a directory containing many NDJSON logs, the
+reference shaper config can be applied to all the records they contain and
+output them all in a single binary [ZNG](../docs/formats/zng.md) file as
+follows:
+
+```
+zq -I shaper.zed *.log > /tmp/all.zng
+```
+
+If you wish to apply the shaping configuration and then perform additional
+operations on the richly-typed records, the Zed query on the command line
+should begin with a `|`, as this appends it to the pipeline at the bottom of
+the shaper from the included file.
+
+For example, to count Zeek `conn` records into CIDR-based buckets based on
+originating IP address:
+
+```
+zq -I shaper.zed -f table '| count() by network_of(id.orig_h) | sort -r' conn.log
+```
+
+[zed/2584](https://github.com/brimdata/zed/issues/2584) tracks a planned
+improvement for this use of `zq -I`.
+
+If you intend to frequently shape the same NDJSON data, you may want to create
+an [alias](https://tldp.org/LDP/abs/html/aliases.html) in your
+shell to always invoke `zq` with the necessary `-I` flag pointing to the path
+of your finalized shaper. [zed/1059](https://github.com/brimdata/zed/issues/1059)
+tracks a planned enhancement to persist such settings within Zed itself rather
+than relying on external mechanisms such as shell aliases.
+
+# Zeek Version/Configuration
+
+The fields and data types in the reference `shaper.zed` reflect the default
+NDJSON-format logs output by the Zeek version referenced in the comments at the
+top of that file. This configuration has been revisited periodically as new
+Zeek versions have been released and in that time we've only seen a couple new
+fields appear and none have been removed. Because of this, we expect the
+shaper configuration should be usable as-is for Zeek releases older than the
```

Comment:
```suggestion
top of that file. They have been revisited periodically as new
Zeek versions have been released and in that time we've only seen a couple new
fields appear and none have been removed. Because of this, we expect the
shaper should be usable as-is for Zeek releases older than the
```